Fix #13262 admin API renderer classes when rest_framework not installed

vijay-x-Raj opened the pull request:
## Fix admin API renderer classes when rest_framework not installed

Fixes #13262

### Problem
The admin API endpoints (e.g., `/admin/api/main/pages/`) were causing template/static file errors when accessed directly in the browser if the `rest_framework` app wasn't installed. This happened because the `PagesAdminAPIViewSet` inherited `renderer_classes` from the base class which included `BrowsableAPIRenderer`, but this renderer requires DRF templates and static files.

### Solution
- Override the `renderer_classes` property in `PagesAdminAPIViewSet` to dynamically determine which renderers to use based on installed apps
- Only use `JSONRenderer` when `rest_framework` is not installed to avoid template/static file errors
- Keep `BrowsableAPIRenderer` when `rest_framework` is installed for better developer experience
- Add comprehensive tests to verify the renderer behavior works correctly

### Changes
- Modified `wagtail/admin/api/views.py` to add dynamic `renderer_classes` property
- Added `wagtail/admin/tests/api/test_renderer_classes.py` with tests for different scenarios

### Testing
- ✅ All existing admin API tests continue to pass
- ✅ New tests verify correct renderer selection based on app installation
- ✅ API always returns JSON when `rest_framework` is not installed
- ✅ API continues to support both JSON and HTML when `rest_framework` is installed

### Implementation Details
The solution uses `django.apps.apps.is_installed("rest_framework")` to dynamically determine which renderers to include. This approach:
- Maintains backward compatibility
- Provides the best user experience in both scenarios
- Avoids the template/static file errors mentioned in the issue
- Preserves the browsable API functionality when DRF is properly installed

laymonage commented on wagtail/admin/api/views.py:
```diff
@@ -26,6 +28,23 @@ class PagesAdminAPIViewSet(PagesAPIViewSet):
     base_serializer_class = AdminPageSerializer
     authentication_classes = [SessionAuthentication]
 
+    @property
+    def renderer_classes(self):
```
Can we put this in https://github.com/wagtail/wagtail/blob/bb8800927c74207fe7de9f4efca8a8961e2f95ab/wagtail/api/v2/views.py#L37-L38 instead as per https://github.com/wagtail/wagtail/issues/13265#issuecomment-3107821235?

laymonage commented on wagtail/admin/tests/api/test_renderer_classes.py:
```diff
@@ -0,0 +1,70 @@
+        self.assertIn("meta", content)
+        self.assertIn("items", content)
+
+    def test_api_response_returns_json_with_json_accept_header(self):
```
Can we have another test that verifies the API can still return HTML with the right `Accept` header?

laymonage reviewed (commented):
Thanks!

vijay-x-Raj commented:
Hi @laymonage 
I've addressed your feedback:

- moved `renderer_classes` to `BaseAPIViewSet` in `wagtail/api/v2/views.py`
- added HTML rendering

thanks for the feedback

laymonage commented on wagtail/api/v2/tests/test_renderer_classes.py:
```diff
@@ -0,0 +1,88 @@
+        self.assertIn("meta", content)
+        self.assertIn("items", content)
+
+    def test_api_response_returns_html_with_html_accept_header(self):
+        """Test that API returns HTML when HTML is explicitly requested via Accept header."""
+        response = self.client.get(
+            reverse("wagtailapi_v2:pages:listing"), HTTP_ACCEPT="text/html"
+        )
+
+        self.assertEqual(response.status_code, 200)
+        self.assertEqual(response["Content-Type"], "text/html; charset=utf-8")
+
+        # Should contain HTML content
+        content = response.content.decode("UTF-8")
+        self.assertIn("<html", content)
+        self.assertIn("</html>", content)
+
+    def test_api_response_returns_html_with_browser_accept_header(self):
```
Turns out we can't have these tests due to the test mechanism added in #12251... I'll remove them for now and raise a separate PR to change that mechanism.
